aiida/tools/visualization/graph.py

```diff
@@ -571,48 +645,68 @@ def recurse_descendants(
-                    self.add_incoming(node, link_types=link_types, annotate_links=annotate_links)
-
-            # ensure the same path isn't traversed multiple times
-            leaf_nodes = []
-            for new_node in new_nodes:
-                if new_node.pk in traversed_pks:
-                    continue
-                leaf_nodes.append(new_node)
-                traversed_pks.append(new_node.pk)
+        # Traverse backward along input_work and input_calc links from all nodes traversed in the previous step
+        # and join the result with the original traversed graph. This includes calculation inputs in the Graph
+        if include_process_inputs:
+            inputs_traversal_rules = self._link_types_to_traversal_rules(
+                link_types=('input_calc', 'input_work'), direction='backward'
+            )
+            traversed_outputs = traverse_graph(
+                traversed_graph['nodes'], max_iterations=1, get_links=True, **inputs_traversal_rules
+            )
+            traversed_graph['nodes'] = traversed_graph['nodes'].union(traversed_outputs['nodes'])
+            traversed_graph['links'] = traversed_graph['links'].union(traversed_outputs['links'])
+
+        # Do one central query for all nodes in the Graph and generate a {id: Node} dictionary
+        traversed_nodes = QueryBuilder().append(
+            Node, filters={'id': {
+                'in': traversed_graph['nodes']
+            }}, project=['id', '*'], tag='node'
+        )
+        traversed_nodes = {query_result[0]: query_result[1] for query_result in traversed_nodes.all()}
+
+        # Pop the origin node and add it to the graph, applying custom styling
+        origin_node = traversed_nodes.pop(origin_pk)
+        self.add_node(origin_node, style_override=origin_style)
+
+        # Add all traversed nodes to the graph with default styling
+        for _, traversed_node in traversed_nodes.items():
+            self.add_node(traversed_node, style_override={})
+
+        # Add the origin node back into traversed nodes so it can be found for adding edges
+        traversed_nodes[origin_pk] = origin_node
+
+        # Add all links to the Graph, using the {id: Node} dictionary for queryless Node retrieval, applying
+        # appropriate styling
+        for link in traversed_graph['links']:
+            source_node = traversed_nodes[link.source_id]
+            target_node = traversed_nodes[link.target_id]
+            link_pair = LinkPair(self._convert_link_types(link.link_type)[0], link.link_label)
+            link_style = self._link_styles(
+                link_pair, add_label=annotate_links in ['label', 'both'], add_type=annotate_links in ['type', 'both']
+            )
+            self.add_edge(source_node, target_node, link_pair, style=link_style)
 
     def recurse_ancestors(
         self,
         origin,
         depth=None,
         link_types=(),
         annotate_links=False,
-        origin_style=(),
+        origin_style=None,
```

Comment:
I encountered some errors with the original default. I'm not really sure why.

Reply:
The problem is with the call to `add_node` with the empty tuple in `origin_style`, more specifically in these two lines:

```
style = {} if style_override is None else style_override
style.update(self._global_node_style)
```

The error is triggered by the second line, but the problem is actually on the first. If we wanted to keep the default as an empty tuple we could change the condition to 'if not style_override', but I'm hesitant to do this without better understanding how 'add_node' works. As a matter of fact, I don't get why the override is loaded first and the "default" (I understand that is `_global_node_style`) is updated after. Isn't it supposed to first load defaults and then overwrite them if necessary? Is there something I am missing? (I'll summon @chrisjsewell since this seems to be the original implementation)

Reply:
Also, I don't really understand how tuples would work with that algorithm. For now I will change the default to `None` and write in the docstring that `None` and `dict` are the acceptable types (idem `recurse_descendants` for consistency). For now I don't want to put checks that would raise with tuples since I may be missing something and I could be affecting backwards compatibility.

Reply:
OK to also keep this then as is